README
germanium
=========

Germanium is a set of extensions on top of the regular WebDriver API, allowing
for a super easy creation of tests. It's opensource and free.

Simply put Germanium is a Web Testing API that doesn't suck.

Here is some draft documentation: [doc/api/index.adoc](doc/api/index.adoc)

ChangeLog
---------

* 2016-02-11  1.3.7  Renamed README so it should appear in Pypy hopefully.
* 2016-02-11  1.3.6  Added multikey typing: `type_keys(g, '<ctrl-left*3>')`
* 2016-02-03  1.3.5  Added `germanium.js()`. Added documentation.
* 2016-01-28  1.3.4  *BugFix* Detect if the node is an element, by nodeType and not instanceof. (Fixes Chrome issues)
* 2016-01-26  1.3.3  *BugFix* Fixed the `type_keys` implementation for IE.
* 2016-01-25  1.3.2  *BugFix* `S` locator doesn't throw when is not finding elements.
* 2016-01-25  1.3.1  *BugFix* Release script.
* 2016-01-25  1.3.0  Added `S` super locator.
* 2016-01-25  1.2.0  Added `wait` utility function. Added a bunch of tests.
* 2015-12-03  1.1.1  Fixed returning `dict` object instead of `WebElement` under python 3.4.
* 2015-11-30  1.1.0  Added `type_keys` API.
* 2015-11-30  1.0.1  PIP deployment fix for python 3.4.



CHANGELOG.md
ChangeLog
---------

* 2017-01-03  1.10.5
    - Updated chrome drivers to 2.27.

* 2016-12-14  1.10.4
    - `outside` filtering for elements added. You can now filter for elements that are `outside` specific elements.
    - `Text()` now accepts both `exact` and `trim` matches.
    - Updated drivers for Chrome to 2.26.

    _Improvements_

    This version is fully tested on a Germanium provisioned Selenium grid: [https://github.com/germaniumhq/germanium-docker](https://github.com/germaniumhq/germanium-docker).

* 2016-11-17  1.10.3
    - `while_not` regression fix.

* 2016-11-14  1.10.2
    - `while_not` has now its results resolving as well, so lambdas are recursively resolved.

* 2016-11-06  1.10.1
    - Updated drivers for Chrome.

* 2016-10-24  1.10.0
    - Edge support (beta).

    _Bugfixes_

    - Performance improvements + better stability on `is_displayed` for multiple elements.
    - `child-nodes` is now also a minified script (Germanium 2.0)

* 2016-10-15  1.9.10
    _Bugfixes_

    - `requireWindowFocus` by default on IE profiles, so typing keys in IE11 it's not abysmally slow.
    - Updated marionette drivers (for Firefox)
    - Refactored some tests towards Germanium 2.0

* 2016-10-03  1.9.9
    _Bugfixes_

    - Allow using a `WebElement` instance for filtering.

* 2016-09-13  1.9.8
    - `waited` function call. You can write now `click(waited(Text(..)))` that will also wait.
    - Made the static selector allow also lists of elements for locators.

    _Bugfixes_

    - Improved `S`/`create_locator` to deal with lists of elements.
    - Updated the chrome driver to 2.24.

* 2016-08-23  1.9.7
    - Removed Firefox support back to 47.0.1. Seems Firefox abandoned Selenium 2, and dived straight into 3-beta.
    - `type_keys` has now a `delay` parameter, to allow a delay between key presses.

    _Bugfixes_

    - Fixed bug in positional filtering where elements bordering on the left/right, or top-bottom, were marked as overlapping.

* 2016-08-19  1.9.6
    - Firefox 48.0.1 support.

* 2016-08-15  1.9.5
    - InputText allows as `type` all the HTML5 types that generate inputs that allow typing.

* 2016-08-13  1.9.4
    - Added `left`, `right`, `top`, `bottom` functions to get the values.
    - Added `germanium.version` package to get the version numbers.

    _Bugfixes_
    - Match also submit buttons on the `Button` selector: `input[type='submit']`
    - InputText searches for the placeholder if it's present.
    - Fixed a timing issue in alert closing on IE, where alerts would still be reported as present by selenium, even after closing them.

* 2016-08-09  1.9.3
    - Added `Box`, and `Point` support, to able to easily click at custom point positions. E.g.
```py
click(Box(Css('div.title')).bottom_right(2, 10))
```

    _Bugfixes_
    - Use `outline` instead of `background-color` to do highlights.
    - Install parcellite on the docker instance, to have a clipboard across browser instances.
    - Update the drivers for chrome, and firefox. On docker we still use Firefox 47.0.1, since 48.0 has yet again broken selenium support.
    - Fixed the `get_text()` on IE9 returning sometimes a different `string` than the Chrome/FF conterpart.
    - Recursively resolve callables inside `wait()` calls.
    - inside/contains filters can now filter against invisible elements as well. `assert Css("p").containing("br").exists()`.
    - show the current version in the docker log.
    - Alert handling fixes.

* 2016-08-04  1.9.2
    - Added `Window` selector, and `WindowLocator`, to easily find windows.

    _Bugfixes_
    - All the default selectors are returning relative XPaths.
    - Cleaned up legacy code.
    - Deal with `None` returned sometimes by Selenium in relative CSS searches.
    - InsideFilterLocator can accept now locators that return `None` instead of empty list.

* 2016-08-01  1.9.1
    - Added `use_window` to permit selecting the active window that we're working with.

    _Bugfixes_

    - Added `inside`/`containing`/`containing_all` checks for empty nodes, so `Element("div").containing(Text("missing")) will return empty, instead of all the divs.
    - Throw meaningful error messages if the property is not defined on the Germanium/WebDriver instance.

* 2016-07-29  1.9.0
    - Completely rewrote `inside`/`containing`/`without_elements` as filters. This extends from only XPath selectors that can be used as filters to _any_ kind of selector (CSS, JS, or custom `AbstractSelector` class).
    - Added `containing_all` to match elements that contain all the given selectors.
    - Documentation updates.

    _Bugfixes_
    - Positional filtering sorting fixes, now preffering elements that overlap.
    - XPath `Element` indexes, if they happen also on nested elements using `containing`, works as expected.

* 2016-07-25  1.8.3
    _Bugfixes_

    - Sort the elements vertically/horizontally on the positional filtering.

* 2016-07-19  1.8.2
    - Rewrote the positional filtering. Blazing fast.

* 2016-07-12  1.8.1
    - Updated documentation.
    - Python upgrades: 3.5.2, 3.4.5, 2.7.12
    - Allow space separated CSS classes in the `Element` locator.

    _Bugfixes_

    - Don't die when WebDriver returns `None` instead of empty list.

* 2016-07-06  1.8.0
    - Firefox 47.0.1 support.
    - selenium 2.52.6 support.
    - Allow instantiation of Firefox with or without marionette.
    - Use drivers that are already packed in the `germaniumdrivers`.

* 2016-06-14  1.7.14
    - Firefox 46 support.
    - selenium 2.52.4 support.
    - `file_select()` renamed to `select_file()`.
    - Documentation update.

* 2016-06-10  1.7.13
    - New function `file_select()` allows selecting the file to be uploaded into a file input.
    - `InputFile` selector added.

* 2016-06-06  1.7.12
    - New function `parent_node()` allows getting the parent node of an element (selector).
    - New function `child_nodes(only_elements=True)` allows getting the child nodes/elements of a parent element (selector).

germanium/driver.py
import pkg_resources

from .locators import CssLocator, XPathLocator, JsLocator, AlertLocator
from .iframe_selector import DefaultIFrameSelector, CallableIFrameSelector
from .create_locator import create_locator
from .impl._alert_exists import _alert_exists, allow_alert

from .impl import wait
from .impl._workaround import workaround
from .wa.ie_eq_8_load_support_scripts import _is_ie_8, _ie_8_load_support_scripts

from selenium.webdriver.remote.webelement import WebElement


class JavaScriptException(Exception):
    def __init__(self, name, message):
        self.name = name
        self.message = message

    def __str__(self):
        return "JavaScriptException: %s - %s" % (self.name, self.message)


class GermaniumDriver(object):
    """
    A Germanium extension to top of WebDriver
    """
    def __init__(self,
                 web_driver,
                 iframe_selector=DefaultIFrameSelector(),
                 screenshot_folder="screenshots",
                 scripts=list()):

        self.web_driver = web_driver
        self._screenshot_folder = screenshot_folder
        self._current_iframe = None
        self._scripts_to_load = scripts

        # set the iframe selector using the property
        self.iframe_selector = iframe_selector

        self.locator_map = {
            "xpath": XPathLocator,
            "css": CssLocator,
            "js": JsLocator,
            "alert": AlertLocator
        }

        self.select_iframe("default")

    def S(self, locator, strategy='detect'):
        """ Finds an element by the given locator.
        :param locator:
        :param strategy:
        """
        return create_locator(self, locator, strategy)

    def get(self, url, *args, **kwargs):
        @allow_alert(self)
        def get_url():
            self.web_driver.get(url, *args, **kwargs)

        get_url()
        self.wait_for_page_to_load()

    def reload_page(self):
        """
        Reloads the page via JS, and waits for it to load.
        """
        self.js('document.location.reload();')
        self.wait_for_page_to_load()

    def execute_script(self, script, *args, **kwargs):
        """ deprecated js """
        return self.js(script, *args, **kwargs)

    #
    # This executes a script taking care of actually catching and rethrowing correctly
    # JavaScript exceptions.
    #
    # It takes special care for returning web elements directly since if they are exported
    # using a map, under python-3.4 webdriver gets dizzy, and returns the elements as "dict"
    #
    def js(self, script, *args, **kwargs):
        try:
            """
            Execute the script, and also display it on the console for debug purposes.
            """
            wrapper_script = """try {
                function __isElement(n) {
                    return n && n.nodeType && n.ownerDocument
                }

                function __isElementList(l) {
                    if (l && typeof l['length'] != 'undefined') {
                        for (var __i = 0; __i < l.length; __i++) {
                            if (!__isElement(l[__i])) {
                                return false;
                            }
                        }

                        return true;
                    }
                }

                var result = {
                    data : (function() {
                        %s
                    }).apply(this, arguments),
                    status : "SUCCESS"
                };

                return __isElement(result.data) || __isElementList(result.data) ?
                       result.data :
                       result;
            } catch (e) {
                return {  // return the exception information in case of failure.
                    status : "FAILURE",
                    name : e.name,
                    message : e.message
                };
            }
            """ % script

            eval_script = wrapper_script

            @allow_alert(self)
            def execute_js():
                return self.web_driver.execute_script(eval_script, *args, **kwargs)

            response = execute_js()

            if isinstance(response, WebElement) or isinstance(response, list):
                return response

            if not isinstance(response, dict):
                return response

            if response['status'] == 'SUCCESS':
                return response['data']
            else:
                raise JavaScriptException( response['name'], response['message'] )
        except Exception as e:
            print(wrapper_script)
            print("Failure executing script: %s, error: %s" % (script, e))
            raise e

    def wait_for_page_to_load(self, timeout=30):
        """
        Wait for the page to load.
        """
        self.select_iframe("default")

        # Checking for alerts must happen first, because Firefox can't
        # execute JS if there is an alert present, and it closes the alert
        # if it has a JS eval error.
        wait(lambda: _alert_exists(self),
             lambda: self.js("""return "complete" == document["readyState"]"""),
             timeout=timeout)

        if _alert_exists(self):
            print("WARNING: Since an alert was present, wait_for_page_to_load "
                  "exited prematurely, and the support scripts were not loaded. "
                  "You need to call `get_germanium().load_support_scripts()` "
                  "manually.")
        else:
            self.load_support_scripts()

    def wait_for_javascript(self, script, timeout = 60):
        """
        Executes a script every 400 milliseconds until it returns true. If it goes more than timeout seconds, then this
        function throws an exception. If the given script throws an exception, it's assumed that it returned false.
        """
        wrapper_script = """
            try {
                %s // original script
            } catch (e) {
                console.log(e);
                return false;
            }
        """ % script

        wait(lambda: self.js(wrapper_script), timeout=timeout)

    @workaround(_is_ie_8, _ie_8_load_support_scripts)
    def load_support_scripts(self):
        """
        Since the support scripts are quite big, they should be loaded independently.
        """
        if self.js("return !window.__GERMANIUM_EXTENSIONS_LOADED"):
            for script_name in self._scripts_to_load:
                self.load_script(script_name)

            self.load_script('germanium-extensions-loaded.js')

    def load_script(self, script_name):
        """
        Load an external script into the current window context.
        """
        script = pkg_resources.resource_string(__name__, script_name)
        if type(script) != 'str':  # it is bytes
            script = script.decode('utf-8')

        self.js(script)

    def take_screenshot(self, name, *args, **kwargs):
        """
        Takes a screenshot of the current browser.
        :param name:
        :return:
        """
        path = self._screenshot_folder + "/" + name + ".png"
        self.web_driver.get_screenshot_as_file(path, *args, **kwargs)
        print("Screenshot saved as " + path)

    def select_iframe(self, iframe_name):
        """
        Selects the iframe, only if the iframe is different
        :param iframe_name:
        :return:
        """
        if iframe_name != self._current_iframe:
            self._current_iframe = self._iframe_selector.select_iframe(self, iframe_name)

        if not self._current_iframe:
            self._current_iframe = iframe_name

        return self._current_iframe

    @property
    def iframe_selector(self):
        return self._iframe_selector

    @iframe_selector.setter
    def iframe_selector(self, value):
        if hasattr(value, '__call__'):
            value=CallableIFrameSelector(value)

        self._iframe_selector = value

    def __getattr__(self, item):
        """
        Delegate all the attributes that are missing to the web_driver.
        :param item:
        :return:
        """
        if "current_iframe" == item:
            return self._current_iframe

        if not hasattr(self.web_driver, item):
            raise Exception("Attribute `%s` doesn't exist on either the Germanium instance, nor "
                            "on the WebDriver instance." % item)

        return getattr(self.web_driver, item)
